exonum/src/helpers/mod.rs

```diff
@@ -77,3 +79,62 @@ pub fn generate_testnet_config(count: u16, start_port: u16) -> Vec<NodeConfig> {
             thread_pool_size: Default::default(),
         }).collect::<Vec<_>>()
 }
+
+/// This routine is adapted from the *old* Path's `path_relative_from`
+/// function, which works differently from the new `relative_from` function.
+/// In particular, this handles the case on unix where both paths are
+/// absolute but with only the root as the common directory.
+///
+/// @see https://github.com/rust-lang/rust/blob/e1d0de82cc40b666b88d4a6d2c9dcbc81d7ed27f/src/librustc_back/rpath.rs#L116-L158
+pub fn path_relative_from(path: impl AsRef<Path>, base: impl AsRef<Path>) -> Option<PathBuf> {
+    let path = path.as_ref();
+    let base = base.as_ref();
+
+    if path.is_absolute() != base.is_absolute() {
+        if path.is_absolute() {
+            Some(PathBuf::from(path))
+        } else {
+            None
+        }
+    } else {
+        let mut ita = path.components();
+        let mut itb = base.components();
+        let mut comps: Vec<Component> = vec![];
+        loop {
+            match (ita.next(), itb.next()) {
+                (None, None) => break,
+                (Some(a), None) => {
+                    comps.push(a);
+                    comps.extend(ita.by_ref());
+                    break;
+                }
+                (None, _) => comps.push(Component::ParentDir),
+                (Some(a), Some(b)) if comps.is_empty() && a == b => (),
+                (Some(a), Some(b)) if b == Component::CurDir => comps.push(a),
+                (Some(_), Some(b)) if b == Component::ParentDir => return None,
+                (Some(a), Some(_)) => {
+                    comps.push(Component::ParentDir);
+                    for _ in itb {
+                        comps.push(Component::ParentDir);
+                    }
+                    comps.push(a);
+                    comps.extend(ita.by_ref());
+                    break;
+                }
+            }
+        }
+        Some(comps.iter().map(|c| c.as_os_str()).collect())
+    }
+}
+
+#[test]
+fn test_path_relative_from() {
+    let baz: PathBuf = "/foo/bar/baz".into();
+    let bar: PathBuf = "/foo/bar".into();
+    let quux: PathBuf = "/foo/bar/quux".into();
+    assert_eq!(path_relative_from(&bar, &baz), Some("../".into()));
+    assert_eq!(path_relative_from(&baz, &bar), Some("baz".into()));
+    assert_eq!(path_relative_from(&quux, &baz), Some("../quux".into()));
+    assert_eq!(path_relative_from(&baz, &quux), Some("../baz".into()));
+    assert_eq!(path_relative_from(&bar, &quux), Some("../".into()));
```

Comment:
I'd like to see more edge cases in tests for a utility function which may be reused in surprising places later.

- relative paths (as both arguments)
- special path components (`.` and `..`)
- trying to escape the root directory (`/foo/../../../../../../../../../../../bar`)
- any other stupid things that you may think of (e.g., `("/", "/")`, referring the directory itself, etc.)

Reply:
This function does not work some of this cases properly.

>relative paths

Sort of works.

>special path components
>trying to escape the root directory

it does not "collapse" arguments so it does not work in all cases.

>("/", "/"), referring the directory itself

It will return ""...